### PR TITLE
Scott's proposal

### PR DESCRIPTION
**Pull request by bwesterb**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+The changes in this branch are expiremental, and need review (and agreement on the
+changes to the Sphincs+ definition) before pulling.
+It involves using SHA-512 for the H, T functions (rather than SHA-256) in the L3, L5
+parameter sets.  For robust parameter sets, we switch to SHA-512 as well for the masks
+(again, for L3, L5 parameter sets).
+
```
Comment (bwesterb) [inline README.md:6]: `SHA256SUMS` needs to be updated; can be done by using the output of `python3 vectors.py`

File: sha256-avx2/thash_sha256_simplex8.c [hunk 2]
```diff
@@ -8,6 +8,19 @@
 #include "sha256.h"
 #include "sha256x8.h"
 #include "sha256avx.h"
+#include "sha512x4.h"
```
Comment (bwesterb) [inline sha256-avx2/thash_sha256_simplex8.c:11]: Put inside `#ifdef`.

File: sha256-avx2/sha512x4.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+void sha512_init_frombytes_x4(sha512ctx4x *ctx, const void *s, unsigned long long msglen);
+void sha512_init4x(sha512ctx4x *ctx);
+void sha512_update4x(sha512ctx4x *ctx, 
+                     const void *d0,
+                     const void *d1,
+                     const void *d2,
+                     const void *d3,
+                     unsigned long long len);
+void sha512_final4x(sha512ctx4x *ctx,
+                     __m256i out0[2],
```
Comment (bwesterb) [inline sha256-avx2/sha512x4.h:23]: Why `__m256i` here?

File: sha256-avx2/thash_sha256_simplex8.c [hunk 4]
```diff
@@ -91,3 +117,62 @@ void thashx8(unsigned char *out0,
...
+    );
+    sha512_final4x(&hash,
+        /* out */
+        outbufx4 + 0*SPX_SHA512_OUTPUT_BYTES / sizeof(__m256i),
+        outbufx4 + 1*SPX_SHA512_OUTPUT_BYTES / sizeof(__m256i),
+        outbufx4 + 2*SPX_SHA512_OUTPUT_BYTES / sizeof(__m256i),
+        outbufx4 + 3*SPX_SHA512_OUTPUT_BYTES / sizeof(__m256i)
+    );
+
+    memcpy(out0, outbufx4 + 0*SPX_SHA256_OUTPUT_BYTES / sizeof(__m256i), SPX_N);
```
Comment (bwesterb) [inline sha256-avx2/thash_sha256_simplex8.c:173]: 256 -> 512?

File: sha256-avx2/thash_sha256_robustx8.c [hunk 5]
```diff
@@ -8,6 +8,19 @@
 #include "sha256.h"
 #include "sha256x8.h"
 #include "sha256avx.h"
+#include "sha512x4.h"
```
Comment (bwesterb) [inline sha256-avx2/thash_sha256_robustx8.c:11]: Same

File: sha256-avx2/thash_sha256_robustx8.c [hunk 6]
```diff
@@ -116,3 +142,79 @@ void thashx8(unsigned char *out0,
...
+    );
+    sha512_final4x(&hash,
+        /* out */
+        outbufx4 + 0*SPX_SHA512_OUTPUT_BYTES / sizeof(__m256i),
+        outbufx4 + 1*SPX_SHA512_OUTPUT_BYTES / sizeof(__m256i),
+        outbufx4 + 2*SPX_SHA512_OUTPUT_BYTES / sizeof(__m256i),
+        outbufx4 + 3*SPX_SHA512_OUTPUT_BYTES / sizeof(__m256i)
+    );
+
+    memcpy(out0, outbufx4 + 0*SPX_SHA256_OUTPUT_BYTES / sizeof(__m256i), SPX_N);
```
Comment (bwesterb) [inline sha256-avx2/thash_sha256_robustx8.c:215]: 256->512

File: sha256-avx2/sha512x4.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#define SHA512AVX_H
+#include <stdint.h>
+#include "immintrin.h"
+
+typedef struct SHA512state4x {
+    __m256i s[8];
+    unsigned char msgblocks[4*128];
+    int datalen;
+    unsigned long long msglen;
+} sha512ctx4x;
```
Comment (bwesterb) [inline sha256-avx2/sha512x4.h:11]: We do not need the incremental API: a single function that combines init_from_bytes; update and final will be faster. Not a blocker for the initial PR, but we should create an issue to remember.

File: sha256-avx2/sha512x4.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
+#ifndef SHA512AVX_H
```
Comment (bwesterb) [inline sha256-avx2/sha512x4.h:1]: SPX prefix.

File: ref/context.h [hunk 9]
```diff
@@ -12,6 +12,8 @@ typedef struct {
 #ifdef SPX_SHA256
     // sha256 state that absorbed pub_seed
     uint8_t state_seeded[40];
+    // sha512 state that absorbed pub_seed
+    uint8_t state_seeded_512[72];
```
[the remark this answers is not in the document]
Comment (sfluhrer) [reply ref/context.h:16]: I agree; I originally didn't have a global SPX_DO_SHA512; when I added it, I forgot to wrap this...

File: sha256-avx2/Makefile [hunk 10]
```diff
@@ -1,12 +1,12 @@
-PARAMS = sphincs-sha256-128f
+PARAMS = sphincs-sha256-256f
```
[the remark this answers is not in the document]
Comment (sfluhrer) [reply sha256-avx2/Makefile]: Done

File: sha256-avx2/thash_sha256_robustx8.c (hunk 6 above)
Comment (sfluhrer) [reply sha256-avx2/thash_sha256_robustx8.c]: Corrected

File: sha256-avx2/thash_sha256_simplex8.c (hunk 4 above)
Comment (bwesterb) [reply sha256-avx2/thash_sha256_simplex8.c]: A propos, does the performance anomaly still exist?

**Comment by bwesterb** [comment]
I'm fixing some of the open issues now...

File: sha256-avx2/params/params-sphincs-sha256-128f.h [hunk 11]
```diff
@@ -1 +1,81 @@
-../../ref/params/params-sphincs-sha256-128f.h
\ No newline at end of file
+#ifndef SPX_PARAMS_H
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sha256-avx2/params/params-sphincs-sha256-128f.h:1]: Fixed.

File: README.md (hunk 1 above)
Comment (bwesterb) [reply README.md:6]: Done.

File: ref/context.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (bwesterb) [reply ref/context.h:18]: Done.

**Comment by bwesterb** [comment]
> I'm fixing some of the open issues now...

Done. All yours again.

**Comment by bwesterb** [comment]
There is something wrong with the AVX2 implementation.

**Comment by bwesterb** [comment]
> There is something wrong with the AVX2 implementation: see CI failures.

Two block thashx8 seems off. I've added a separate test. Digging down further.

**Comment by sfluhrer** [comment]
The SHA512x4 implementation passed KATs; however there is something wrong (probably with the H/T implementations); I'm digging into that...

**Comment by bwesterb** [comment]
> The SHA512x4 implementation passed KATs; however there is something wrong (probably with the H/T implementations); I'm digging into that...

You're right — I had a mistake in my SHA512x4 test.

**Comment by bwesterb** [comment]
There were two errors:

1. big/little endian confusion mismatch when loading the seeded state and
2. 512 instead of 1024 initial message length for seeded state.

**Comment by bwesterb** [comment]
I'm ripping out the incremental API now to bring it more in line with the sha256x8 code.

**Comment by bwesterb** [comment]
Ok, there is a new bug. 🤔

**Comment by sfluhrer** [comment]
Thanks!  I had found the second one; I was still searching for the first one…

**Comment by bwesterb** [comment]
Could it be https://github.com/sphincs/sphincsplus/pull/16#discussion_r499538595 ?

**Comment by bwesterb** [comment]
ea515e88fe93b6de09a944961b4b3805a7aeaf6a is good

**Comment by bwesterb** [comment]
6ceec71971a481579497f1ada1452b7ff207955c is bad

**Comment by bwesterb** [comment]
I think we must have introduced some undefined behaviour in the big tree changes. Have a look at https://github.com/sphincs/sphincsplus/tree/scott-example

**Comment by bwesterb** [comment]
The problem was that thashx8 was called with overlapping in- and outputs. I rewrote thashx8 to handle that correctly by reading all inputs before writing the outputs.

File: sha256-avx2/thash_sha256_simplex8.c (hunk 4 above)
Comment (bwesterb) [reply sha256-avx2/thash_sha256_simplex8.c]: I think the write-before-read was the cause of the performance anomaly.

**Comment by bwesterb** [comment]
Running benchmark now...

**Comment by bwesterb** [comment]
```
sphincs-sha256-256s robust using sha256-avx2 Verifying
 old: [6927873, 6843298, 6871251, 7083907, 6783781, 6889098, 6677046, 6955137, 6784146]
 new: [5327553, 5338741, 5400439, 5434803, 5419242, 5285101, 5379558, 5414256, 5446320]

sphincs-sha256-256f robust using sha256-avx2 Verifying
 old: [12739668, 12628918, 12614049, 12538183, 12874776, 12773670, 12739477, 12525684, 12644922]
 new: [10418325, 10172119, 10442601, 10350138, 10291825, 10560637, 10378615, 10344729, 10439925]

sphincs-sha256-256s simple using sha256-avx2 Verifying
 old: [1759081, 1761754, 1795401, 1759480, 1836547, 1802080, 1792893, 1758589, 1797150]
 new: [1566405, 1547265, 1523016, 1527630, 1505596, 1494202, 1597225, 1532227, 1579096]

sphincs-sha256-192s robust using sha256-avx2 Verifying
 old: [2675323, 2687068, 2590639, 2664021, 2618745, 2661766, 2675778, 2634600, 2689240]
 new: [2268543, 2219295, 2245230, 2257942, 2251594, 2245534, 2289592, 2249992, 2255034]

sphincs-sha256-192f simple using sha256-avx2 Generating keypair
 old: [1769552, 2076048, 2154325, 2083595, 1879296, 2394117, 2076063, 1616894, 1847838]
 new: [1389474, 1853738, 1393987, 2083086, 1463712, 2646282, 1686674, 1636324, 2136628]

sphincs-sha256-192s simple using sha256-avx2 Verifying
 old: [1239361, 1270578, 1269084, 1301008, 1255411, 1263400, 1265545, 1236982, 1239874]
 new: [1066125, 1116411, 1127671, 1103394, 1115253, 1093176, 1098060, 1091916, 1119553]

sphincs-sha256-256s robust using ref Signing
 old: [10869980937, 10875982552, 10870917540, 10849895380, 10852806111, 10851569697, 10887634387, 10854326050, 10855682718]
 new: [9524795374, 9502611895, 9520798438, 9531359616, 9512573221, 9578471109, 9508197259, 9513259785, 9514508377]

sphincs-sha256-256f simple using sha256-avx2 Verifying
 old: [3085368, 3106551, 3130029, 3090513, 3074332, 3176268, 3103234, 3111892, 3133215]
 new: [2732206, 2761804, 2742349, 2778700, 2744284, 2722390, 2765377, 2829187, 2750760]

sphincs-sha256-192f robust using sha256-avx2 Verifying
 old: [6381168, 6287451, 6319767, 6272461, 6405157, 6340924, 6436083, 6409963, 6414243]
 new: [5644612, 5574060, 5748381, 5629749, 5686849, 5570068, 5771473, 5737666, 5597533]

sphincs-sha256-256s robust using ref Verifying
 old: [16216486, 16902673, 16646916, 16845136, 16992970, 16706191, 16309096, 15884782, 16121863]
 new: [14493291, 14819455, 14342512, 14592831, 14503371, 14114581, 14526219, 14533839, 14348256]

sphincs-sha256-192f simple using sha256-avx2 Verifying
 old: [2985529, 3008290, 2958645, 2969329, 3017073, 3016627, 2976684, 2941165, 2999386]
 new: [2659773, 2734504, 2665038, 2711161, 2682906, 2672904, 2710524, 2682319, 2704620]

sphincs-sha256-256f robust using ref Verifying
 old: [32905872, 32577441, 32250657, 32499348, 33245836, 32942859, 31975840, 32217865, 31894861]
 new: [29230177, 30580176, 29391870, 29870884, 30266379, 29771620, 30566022, 30286657, 29522310]

sphincs-sha256-256f robust using ref Signing
 old: [1125322087, 1124717994, 1123978314, 1124872398, 1123286545, 1123380648, 1131654120, 1123472038, 1124171425]
 new: [1036154112, 1034997198, 1035733389, 1035818325, 1038361924, 1036381422, 1035951861, 1034720137, 1034650708]

sphincs-sha256-256s robust using sha256-avx2 Signing
 old: [2873422108, 2884935399, 2875243161, 2875306014, 2880132171, 2874629200, 2891778747, 2874210256, 2871361497]
 new: [2654903710, 2650604434, 2652136333, 2656943106, 2661367228, 2651450301, 2651836173, 2655548047, 2649384180]

sphincs-sha256-192s robust using ref Verifying
 old: [7752318, 7915516, 8085021, 7828678, 8101473, 7925992, 7820328, 7647688, 7588039]
 new: [7029562, 7329924, 7242594, 7169962, 7193940, 7378990, 7464022, 7464609, 7296628]

sphincs-sha256-256f robust using sha256-avx2 Signing
 old: [308084183, 307696437, 308230014, 309749541, 308431141, 307737837, 310512510, 307951308, 311863986]
 new: [293025760, 292357899, 291404304, 291023074, 291389203, 292328005, 292900629, 291414349, 294879958]

sphincs-sha256-256f simple using sha256-avx2 Generating keypair
 old: [4908304, 3951097, 4549202, 5006238, 4446946, 5376757, 3961039, 5306102, 4882098]
 new: [4730326, 4542690, 4439670, 4202698, 5079070, 3749901, 4566028, 4541948, 4302138]

sphincs-sha256-192f robust using ref Verifying
 old: [22270587, 21795433, 22325425, 22185604, 21841356, 22069573, 22051914, 22254166, 22186987]
 new: [21145839, 21160801, 21174265, 21274467, 20745427, 21516532, 21221644, 21703266, 21321411]

sphincs-sha256-256s simple using ref Signing
 old: [4004501083, 3996754780, 4007672310, 4026541311, 3996788473, 3994694127, 3995445144, 3999768727, 4003297117]
 new: [3811803502, 3808954572, 3810213993, 3804921127, 3810934354, 3806731086, 3805648194, 3802722246, 3822894849]

sphincs-sha256-192s robust using ref Signing
 old: [8703220606, 8695682794, 8729691325, 8712892950, 8696723544, 8688901561, 8694169941, 8695871919, 8697062934]
 new: [8277571921, 8285171796, 8284513494, 8273293971, 8283291592, 8297218276, 8278729321, 8277374325, 8294034753]

sphincs-sha256-256s simple using ref Verifying
 old: [5467561, 5564625, 5419315, 5484199, 5349787, 5382076, 5434434, 5452357, 5448274]
 new: [5224734, 5141175, 5232348, 5242122, 5279095, 5400009, 5278038, 5276596, 5132568]

sphincs-sha256-192f robust using ref Signing
 old: [387391918, 388344136, 389161968, 387773995, 387901728, 388142700, 391482378, 388244643, 388783539]
 new: [374076834, 374186269, 372507595, 374351104, 373445412, 373273903, 373470966, 373341811, 373126918]

sphincs-sha256-256s robust using ref Generating keypair
 old: [885467551, 884186892, 894803389, 883296939, 883634554, 884267949, 883591288, 883786615, 890479542]
 new: [852578113, 852043080, 851278333, 856986183, 851236599, 850569690, 852099304, 850482001, 850748928]

sphincs-sha256-256f robust using ref Generating keypair
 old: [55140469, 55217292, 55255566, 55683564, 55310305, 55035801, 55322703, 55104154, 55297488]
 new: [53193387, 53135097, 53180743, 53165404, 53104876, 53192445, 53300017, 53158629, 53147001]

sphincs-sha256-256f simple using ref Signing
 old: [404058762, 404680591, 404147353, 404192419, 403926472, 410017617, 403985691, 403840261, 404238964]
 new: [392428303, 392147542, 392291535, 392158101, 392183146, 392690394, 392287005, 392165736, 392716407]

sphincs-sha256-192s simple using ref Signing
 old: [4443099843, 4441824924, 4445802616, 4438214457, 4436418816, 4440297706, 4434517876, 4438247896, 4437674923]
 new: [4312432132, 4313202556, 4314840685, 4315039143, 4312142643, 4309816345, 4306957098, 4311381210, 4312962609]

sphincs-sha256-256f simple using ref Verifying
 old: [11024271, 10527552, 10572844, 10517250, 10598889, 10616305, 11009296, 10537911, 10781686]
 new: [10498168, 10277403, 10452163, 10304347, 10294588, 10467850, 10508512, 10216833, 10715202]

sphincs-sha256-192f simple using ref Signing
 old: [196311858, 195943425, 196958655, 196388109, 196369231, 196039459, 198482455, 196147003, 196262130]
 new: [190712857, 190856095, 190362094, 190991836, 191023939, 191235819, 190882428, 191619754, 190970992]

sphincs-sha256-256f robust using sha256-avx2 Generating keypair
 old: [15783853, 15486439, 15518100, 15644443, 15568762, 15520105, 15692847, 15777057, 15732244]
 new: [15589757, 15156573, 15116422, 15116511, 15075889, 15079924, 15235917, 15051081, 16800597]

sphincs-sha256-192f robust using sha256-avx2 Generating keypair
 old: [4239784, 3321986, 2985855, 3256248, 3261248, 4045798, 2821725, 3820550, 3022734]
 new: [3054326, 3496342, 4229498, 2888482, 4245611, 3169578, 2859598, 3231094, 2743420]

sphincs-sha256-128f robust using sha256-avx2 Generating keypair
 old: [2439598, 3084531, 3099021, 4238384, 2624918, 2292198, 2430102, 1892088, 1900285]
 new: [1944429, 2056750, 1998906, 2166268, 4269308, 2166836, 1957309, 2516030, 2047258]

sphincs-sha256-192s simple using ref Verifying
 old: [3629208, 3724018, 3895665, 3813222, 3873171, 3741735, 3725047, 3807765, 3528331]
 new: [3582822, 3516954, 3435334, 3648046, 3518257, 3680776, 3538888, 3550059, 3535995]

sphincs-sha256-192s robust using sha256-avx2 Generating keypair
 old: [181701298, 179303619, 179899927, 180552697, 179307451, 179371261, 179745769, 179062980, 181734019]
 new: [174857307, 175830892, 175608385, 178246314, 175414593, 179571948, 177214863, 174695989, 174512364]

sphincs-sha256-192f simple using ref Verifying
 old: [10803342, 10330431, 10658572, 10583302, 10443595, 10366687, 10753972, 10661545, 10690266]
 new: [10073614, 10199146, 10334842, 10290543, 10157755, 10094379, 10500214, 10267245, 10207863]

sphincs-sha256-256s simple using sha256-avx2 Signing
 old: [777340747, 777453895, 768760521, 775626823, 766489498, 774011790, 767311159, 772502439, 768422989]
 new: [786172837, 787059937, 787156099, 786832050, 786017823, 788943502, 785058256, 785598861, 786655861]

sphincs-sha256-128f robust using sha256-avx2 Verifying
 old: [4081354, 3888190, 3959331, 3878097, 4034572, 4095105, 4046032, 4014364, 4049911]
 new: [4030552, 3971655, 4034305, 3978096, 4037625, 4078992, 4153314, 4028776, 4030390]

sphincs-sha256-256s robust using sha256-avx2 Generating keypair
 old: [244806799, 248814411, 245115414, 244673700, 245985321, 244547245, 247334157, 244856919, 244853548]
 new: [246211110, 238757776, 239033007, 240454591, 240294589, 239442826, 239370841, 239180896, 239352222]

sphincs-sha256-128f robust using ref Verifying
 old: [14049864, 14177241, 14053795, 14608557, 13784596, 14784784, 14318167, 14182516, 14765958]
 new: [14246538, 14047180, 14235559, 14306190, 13895328, 13494331, 14452438, 14145186, 14297827]

sphincs-sha256-128s robust using ref Verifying
 old: [4935390, 5154466, 4833657, 4988322, 4764858, 4892263, 4794775, 4726131, 5191927]
 new: [4858926, 5122561, 5023659, 4892184, 4931107, 4635609, 5301274, 4996507, 4997034]

sphincs-sha256-192s robust using ref Generating keypair
 old: [940345581, 939164418, 940790172, 940336663, 939958296, 940120912, 938305573, 941476617, 939290284]
 new: [921979291, 922516510, 922963947, 921567136, 922513980, 922680663, 922127370, 923267592, 928636503]

sphincs-sha256-128f simple using ref Generating keypair
 old: [6403654, 5473688, 5873112, 5379628, 5488127, 5104762, 5547620, 5801535, 6015642]
 new: [5482387, 5667184, 5564589, 5749646, 6014362, 5093509, 5021014, 5393877, 5673584]

sphincs-sha256-192f robust using ref Generating keypair
 old: [14787126, 14812638, 14790946, 14769721, 14897652, 14784214, 14812677, 14990491, 14995957]
 new: [14536609, 14597973, 14569254, 14592226, 14558775, 14620023, 14686881, 14621661, 14572908]

sphincs-sha256-192s simple using sha256-avx2 Generating keypair
 old: [94391425, 91012440, 90627441, 90693762, 90727722, 91279713, 92576581, 92604255, 90748398]
 new: [89834419, 89281539, 89771890, 89498815, 89503026, 89203428, 90192738, 89268039, 89889244]

sphincs-sha256-192f simple using sha256-avx2 Signing
 old: [39296845, 39020814, 39022270, 39276504, 39158610, 39048801, 39182658, 39128982, 39526924]
 new: [38448472, 38469285, 38633697, 38773383, 38482528, 38904030, 38480196, 38426038, 38510536]

sphincs-sha256-256s simple using sha256-avx2 Generating keypair
 old: [61391868, 61532169, 60348793, 61457542, 60038280, 60534852, 60345289, 60581788, 60495556]
 new: [59552986, 59289241, 59162889, 59326305, 59363602, 59142193, 59305186, 59170252, 59277786]

sphincs-sha256-192s simple using sha256-avx2 Signing
 old: [881295207, 862743790, 855970954, 856807191, 855705523, 860218903, 870129339, 885264885, 855385806]
 new: [897838375, 867409738, 869817390, 870303501, 869897896, 873214551, 872582245, 867157701, 872287570]

sphincs-sha256-128f simple using sha256-avx2 Verifying
 old: [1962819, 1972164, 1932798, 1984290, 1897731, 1993525, 1905900, 1931088, 1970449]
 new: [1872271, 1907569, 1940455, 1930864, 1889316, 1920097, 1977363, 1936612, 1918144]

sphincs-sha256-128s simple using sha256-avx2 Verifying
 old: [777592, 730131, 775389, 739180, 733920, 755265, 759217, 723543, 742824]
 new: [749424, 754962, 747784, 749206, 735318, 750045, 777213, 748426, 732156]

sphincs-sha256-256s simple using ref Generating keypair
 old: [309374833, 312242119, 310875451, 311077758, 309657282, 310509963, 310178592, 310789197, 309855423]
 new: [306251406, 306667450, 306272469, 306384273, 306032268, 306295123, 306392169, 306296659, 306328374]

sphincs-sha256-128f simple using ref Verifying
 old: [7091788, 6859837, 6686901, 6973720, 6908485, 6893646, 7059547, 7040295, 7087978]
 new: [7198050, 6977272, 7033315, 7040755, 7452084, 6929325, 6748471, 6794080, 6962236]

sphincs-sha256-192f robust using sha256-avx2 Signing
 old: [75177826, 74728716, 74929924, 75063291, 75466705, 74842323, 74889877, 75527371, 76152415]
 new: [74690595, 74992926, 74439123, 74419060, 74304073, 74375472, 77583456, 74198625, 74080230]

sphincs-sha256-256f simple using ref Generating keypair
 old: [19458432, 19597620, 19472206, 19445181, 19770409, 19577505, 19411866, 19524453, 19532994]
 new: [19264471, 19251207, 19338526, 19300290, 19326337, 19421037, 19276464, 19344198, 19275322]

sphincs-sha256-128f simple using sha256-avx2 Signing
 old: [23804134, 22864530, 23070363, 23089687, 22980225, 22877932, 22936882, 22895776, 23043406]
 new: [23203215, 23561388, 22866961, 22773693, 22774801, 22841568, 22866019, 22731409, 22803162]

sphincs-sha256-128f simple using sha256-avx2 Generating keypair
 old: [1061691, 1684860, 1144670, 1302166, 1614591, 1209062, 970911, 1566070, 1054135]
 new: [981058, 1477068, 1289174, 1041783, 1206294, 1044363, 965638, 1335064, 1014871]

sphincs-sha256-128s simple using ref Verifying
 old: [2504116, 2390575, 2276062, 2275704, 2406990, 2504554, 2275807, 2357695, 2332849]
 new: [2290011, 2638045, 2485464, 2290030, 2518560, 2264187, 2469079, 2420059, 2581978]

sphincs-sha256-192s robust using sha256-avx2 Signing
 old: [1652787519, 1650643833, 1650652585, 1663048771, 1653838513, 1655494054, 1654025454, 1651781680, 1673960398]
 new: [1660350781, 1665241957, 1660857946, 1669889718, 1659255897, 1693532992, 1671351066, 1656781540, 1664370532]

sphincs-sha256-128s robust using sha256-avx2 Verifying
 old: [1494676, 1509052, 1526197, 1505686, 1583742, 1505923, 1554310, 1464055, 1515607]
 new: [1527265, 1519488, 1517358, 1529481, 1527100, 1508203, 1554498, 1458954, 1506513]

sphincs-sha256-192f simple using ref Generating keypair
 old: [9204724, 7401960, 8180739, 8337738, 8535942, 7463008, 8360878, 7720740, 8235362]
 new: [7531057, 8489258, 8013322, 7383033, 7390894, 8498178, 7396335, 7377946, 7609380]

sphincs-sha256-256f simple using sha256-avx2 Signing
 old: [78823782, 78460686, 78605262, 78842977, 78676915, 78897411, 78912496, 79151562, 79031772]
 new: [78467640, 78905362, 78218284, 78331173, 78485457, 78420160, 78322836, 78606153, 78458709]

sphincs-sha256-128f robust using ref Generating keypair
 old: [10020039, 10049950, 10026127, 9993259, 9987589, 10056556, 9995694, 9982561, 9944479]
 new: [10086739, 10148700, 9979105, 9982608, 10003284, 9974883, 9992395, 9967110, 9990696]

sphincs-sha256-192s simple using ref Generating keypair
 old: [468246768, 467504112, 467222769, 467901105, 466837359, 466948854, 466564155, 467081521, 466827004]
 new: [481087216, 466345236, 471199212, 466332373, 466126933, 466451341, 466249255, 465728322, 466539736]

sphincs-sha256-128s robust using sha256-avx2 Generating keypair
 old: [124510227, 121579920, 121465291, 121880431, 120790267, 121052217, 121067776, 121527496, 121253427]
 new: [121288048, 122912712, 121900285, 121512139, 121352898, 122673580, 121093804, 121338585, 120586653]

sphincs-sha256-128f robust using sha256-avx2 Signing
 old: [44789058, 45132697, 46046976, 45227163, 45148998, 44745352, 44786235, 44953507, 45164170]
 new: [45032010, 45069984, 45072993, 44986659, 44826919, 45076323, 46025884, 44945086, 44806536]

sphincs-sha256-128f robust using ref Signing
 old: [228877429, 230732605, 229058560, 229300282, 228995181, 228497820, 229210537, 228477273, 228469600]
 new: [229582032, 228819745, 228959374, 228870025, 229228924, 228926722, 229174396, 229641718, 228764316]

sphincs-sha256-128s robust using sha256-avx2 Signing
 old: [962256531, 923090932, 921244114, 924864043, 917294533, 918535405, 918518697, 922209867, 921577350]
 new: [922107058, 931360041, 924848215, 953265370, 921048250, 930738385, 921238699, 920430946, 916487662]

sphincs-sha256-128s robust using ref Signing
 old: [4755284157, 4741749822, 4751573872, 4749565629, 4740766960, 4750269142, 4749140853, 4744355395, 4744576374]
 new: [4824727582, 4750998861, 4754570257, 4762724673, 4748611147, 4750801560, 4748457259, 4743835108, 4750275577]

sphincs-sha256-128s simple using sha256-avx2 Signing
 old: [476553291, 465053344, 466178289, 465140308, 468025912, 464311758, 464242246, 464617254, 464059792]
 new: [483168867, 466883754, 464380749, 482069998, 467528586, 464405736, 467097703, 464354541, 467998840]

sphincs-sha256-128s simple using sha256-avx2 Generating keypair
 old: [63209878, 61395225, 61396938, 61641640, 61912794, 62743536, 61197213, 61326904, 61390114]
 new: [64263033, 61558947, 63676417, 63382435, 61695352, 61176988, 61681278, 61176229, 61751853]

sphincs-sha256-128s simple using ref Signing
 old: [2418066058, 2416213572, 2415710572, 2418514954, 2419149511, 2415641215, 2418606325, 2416209051, 2417413314]
 new: [2419027570, 2423052186, 2415868921, 2414825004, 2415024088, 2419506871, 2416203841, 2423032311, 2489788618]

sphincs-sha256-128f simple using ref Signing
 old: [116309349, 116150235, 116203762, 116144671, 116185699, 116297986, 116287096, 116908963, 116090434]
 new: [116482963, 116264172, 116055552, 116927661, 116102892, 116189418, 116070699, 116545275, 116100325]

sphincs-sha256-128s robust using ref Generating keypair
 old: [632146095, 632221077, 631943943, 633018115, 632029468, 666600427, 632858404, 631469688, 632633923]
 new: [631792512, 634160680, 633468448, 633806191, 633416176, 631632837, 632290789, 632236758, 632376529]

sphincs-sha256-128s simple using ref Generating keypair
 old: [318359667, 318052254, 318395007, 319470019, 318155836, 318518992, 318332050, 318104736, 318563754]
 new: [318134827, 318033138, 318395823, 318285055, 319208781, 318260058, 318470845, 318001479, 318252873]
```

Merged into the v3.1 PR.

**Pull request merged**